core/src/main/java/hudson/model/AsyncPeriodicWork.java

```diff
@@ -73,8 +124,54 @@ public void run() {
     }
 
     protected StreamTaskListener createListener() {
+        File f = getLogFile();
+        if (!f.getParentFile().isDirectory()) {
+            if (!f.getParentFile().mkdirs()) {
+                logger.log(getErrorLoggingLevel(), "Could not create directory {0}", f.getParentFile());
+            }
+        }
+        if (f.isFile()) {
+            if ((lastRotateMillis + logRotateMillis < System.currentTimeMillis())
+                    || (logRotateSize > 0 && f.length() > logRotateSize)) {
+                lastRotateMillis = System.currentTimeMillis();
+                File p = null;
+                for (int i = 5; i >= 0; i--) {
+                    File o = i == 0 ? f : new File(f.getParentFile(), f.getName() + "." + i);
+                    if (o.isFile()) {
+                        if (p != null && !p.exists()) {
+                            if (!o.renameTo(p)) {
+                                logger.log(getErrorLoggingLevel(), "Could not rotate log files {0} to {1}",
+                                        new Object[]{o, p});
+                            }
+                        } else {
+                            if (!o.delete()) {
+                                logger.log(getErrorLoggingLevel(), "Could not delete log file {0} to enable rotation",
+                                        o);
+                            }
+                        }
+                    }
+                    p = o;
+                }
+            }
+        } else {
+            lastRotateMillis = System.currentTimeMillis();
+            // migrate old log files the first time we start-up
+            File oldFile = new File(Jenkins.getActiveInstance().getRootDir(), f.getName());
+            if (oldFile.isFile()) {
+                File newFile = new File(f.getParentFile(), f.getName() + ".1");
+                if (!newFile.isFile()) {
+                    // if there has never been rotation then this is the first time
+                    if (oldFile.renameTo(newFile)) {
+                        logger.log(getNormalLoggingLevel(), "Moved {0} to {1}.1", new Object[]{oldFile, f});
+                    } else {
+                        logger.log(getErrorLoggingLevel(), "Could not move {0} to {1}.1",
+                                new Object[]{oldFile, f});
+                    }
+                }
+            }
+        }
```

Comment:
`AsyncPeriodicWork.createListener` and `AsyncAperiodicWork.createListener` look to be exactly the same. I don't mind a bit of duplication, but in this case (if I'm right and they are the same) I'd prefer to not see it because (as @rsandell pointed out) the code is not the easiest to read first time (nondescript variable names ... magic number `5` etc).